webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.js

```diff
@@ -0,0 +1,56 @@
+import Immutable from 'seamless-immutable';
+import { translate as __ } from '../../../react_app/common/I18n';
+import {
+  SET_MODAL_OPEN,
+  SET_MODAL_CLOSED,
+  ADD_MODAL,
+} from './ForemanModalConstants';
+
+const initialState = Immutable({});
+
+// Modals state has id as key and open state as value:
+// { myModal: {open: true} }
+// Since keys cannot be duplicated, we avoid creating duplicate modals in this way.
+
+export default (state = initialState, action) => {
+  switch (action.type) {
+    case SET_MODAL_OPEN: {
+      if (!state[action.payload.id]) {
+        throw new Error(
+          __(
+            `SET_MODAL_OPEN error: Modal with id '${action.payload.id}' does not exist`
+          )
+        );
+      }
```

Comment:
how about moving this error throwing to the action?

Reply:
@amirfefer Good idea, will update

Reply:
[update: see comment](https://github.com/theforeman/foreman/pull/7105#discussion_r348769976)

Reply:
How about error handling via redux? which means the correlated action should throw the error and as a result should dispatch an action that changes the state respectively

Reply:
@amirfefer After moving the error handling to Actions instead of Reducers, I'm still thinking Actions isn't the right place for it either.  After making the code change and thinking about it for a bit, I actually think middleware would be the correct place.  One totally legitimate job for middleware is to filter actions and conditionally dispatch them.  (And you'd have access to the current store state easily as well.)

But anyway, I put the error handling in action creators for now since that feels less incorrect than reducers.  Would love to get thoughts from @sharvit also.

Reply:
Errors in javascript is a huge topic :)

When an Error gets `throw`n, the script will stop executing unless you `catch` it. And, it does not matter if you `throw` an Error manually, it is `throw`n by a library, or by the runtime itself (node or the browser).
Therefore, not catching errors is a bad practice since it might/will break the application in an unpredictable way.

IMO there are 2 types of errors
1. **Predicted errors** - The **user** did something wrong and the UI should respond and tell the **user** about it.
A form validation error is an example.

2. **Unpredicted errors** - The **developer** did something wrong and it can lead to unpredictable behavior.
We cannot respond to those errors but we do want to tell the **developer** about what went wrong using logs and it can be nice to keep to UI functioning :)
An example of it is the current error, the **developer** tries to open a model which is not exists.

**IMO it is a good practice having a middleware catching unpredicted errors caused by actions and log them so the developer will have clue about what went wrong.**

BTW, doesn't the `redux-logger` middleware we use in `development` shouldn't catch and log errors in actions by default?
Does it make sense to log those **developer** related errors in production?